### PR TITLE
[FLINK-33264][table] Support source parallelism setting for DataGen connector

### PR DESCRIPTION
**Pull request by X-czh**
## What is the purpose of the change

Support source parallelism setting for DataGen connector.

## Brief change log

- Add SOURCE_PARALLELISM into DataGenConnectorOptions and pass it to planner
- Update docs, introducing scan.parallelism

## Verifying this change

Added UT to verify that the parallelism info is correctly passed to planner.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no) no
  - The serializers: (yes / no / don't know) no
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know) no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know) no
  - The S3 file system connector: (yes / no / don't know) no

## Documentation

  - Does this pull request introduce a new feature? (yes / no) yes
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented) docs

### CODE REVIEW

**Comment by X-czh** [comment]
@libenchao Could you help review it after #24128 is merged?

**Comment by libenchao** [comment]
> @libenchao Could you help review it after #24128 is merged?

I assigned it to myself.

**Comment by X-czh** [comment]
@flinkbot run azure

File: docs/content/docs/connectors/table/datagen.md
Comment (libenchao) [inline docs/content/docs/connectors/table/datagen.md:277]: Actually it's defined as `int` in the code.

File: flink-table/flink-table-api-java-bridge/src/test/java/org/apache/flink/table/factories/DataGenTableSourceFactoryTest.java
Comment (libenchao) [inline flink-table/flink-table-api-java-bridge/src/test/java/org/apache/flink/table/factories/DataGenTableSourceFactoryTest.java:592]: In FLINK-33263, we've introduced a way to verify that the transformation, could you also add a test based on that?

**Review by libenchao — COMMENTED** [review]
@X-czh FLINK-33263 has been merged, could you rebase to the master and we can start to review after that.

File: flink-table/flink-table-api-java-bridge/src/test/java/org/apache/flink/table/factories/DataGenTableSourceFactoryTest.java
Comment (X-czh) [reply flink-table/flink-table-api-java-bridge/src/test/java/org/apache/flink/table/factories/DataGenTableSourceFactoryTest.java:592]: I've already included an IT case in `DataGeneratorConnectorITCase#testWithParallelism` where the physical execution plan is checked, would that be enough?

File: docs/content/docs/connectors/table/datagen.md
Comment (X-czh) [reply docs/content/docs/connectors/table/datagen.md:277]: Thanks for catching this. I'll fix that

Comment (X-czh) [reply docs/content/docs/connectors/table/datagen.md]: Fixed in the new commit

**Review by libenchao — APPROVED** [review]
@X-czh Thanks for the updating, looks good, will merge after the CI passes.

**Comment by XComp** [comment]
@X-czh thanks for discovering the compilation error that was merged into `master`. Could you merge the hotfix commit separately from the bigger change of this PR? The hotfix commit doesn't need to wait for CI to pass and would enable `master` to be stable again.

**Comment by X-czh** [comment]
@XComp Thanks for the notice. We've already prepared another separate PR: #24200 on it.

**Comment by XComp** [comment]
Ah I see, thanks for the pointer. I approved and merge the PR.

**Pull request closed without merging**
